experimental/split_mlir/src/iree/split_mlir/OutlineFunctionsPassImpl.h

```diff
@@ -0,0 +1,285 @@
+        auto insertionResult = operandsSet.insert(operand);
+        if (insertionResult.second) {
+          operands.push_back(operand);
+        }
+      }
+    }
+    for (OpResult result : op.getResults()) {
+      for (OpOperand operand : result.getUsers()) {
+        if (!opsSet.contains(operand.getOwner())) {
+          auto insertionResult = resultsSet.insert(result);
+          if (insertionResult.second) {
+            results.push_back(result);
+          }
+          break;
+        }
+      }
+    }
+  }
+  return {operands, results};
+}
+
+template <typename ValueIt>
+void replaceValueUsesWithNewBlockArguments(ValueIt valuesBegin,
+                                           ValueIt valuesEnd, Block& block) {
+  for (ValueIt valIt = valuesBegin; valIt != valuesEnd; ++valIt) {
+    block.addArgument(valIt->getType(), valIt->getLoc());
+    BlockArgument& blockArg = block.getArguments().back();
+    valIt->replaceUsesWithIf(blockArg, [&block](OpOperand& operand) {
+      return operand.getOwner()->getBlock() == &block;
+    });
+  }
+}
+
+void addBlockReturn(Block& block, ValueRange operands, OpBuilder& builder) {
+  func::ReturnOp returnOp =
+      builder.create<func::ReturnOp>(builder.getUnknownLoc(), operands);
+  block.push_back(returnOp);
+}
+
+void moveOpsIntoBlock(iterator_range<Block::iterator> opRange, Block& block) {
+  // Put ops into another container because opRange will be invalidated during
+  // removal.
+  SmallVector<Operation*, 64> ops;
+  std::transform(opRange.begin(), opRange.end(), std::back_inserter(ops),
+                 [](Operation& op) { return &op; });
+  for (Operation* op : ops) {
+    op->moveBefore(&block, block.end());
+  }
+}
+
+void moveBlock(Region& srcRegion, Region& destRegion,
+               Region::iterator srcBlockIt, Region::iterator destBlockIt) {
+  Block* block = srcRegion.getBlocks().remove(srcBlockIt);
+  destRegion.getBlocks().insert(destBlockIt, block);
+}
+
+bool isAncestorOfBlock(Operation* op, Block* block) {
+  // Walk up the operation hierarchy and check each block.
+  while (op != nullptr) {
+    if (op->getBlock() == block) {
+      return true;
+    }
+    op = op->getParentOp();
+  }
+  return false;
+}
+
+template <typename OriginalOpResultsIt, typename NewOpResultsIt>
+void substititeUses(OriginalOpResultsIt originalBegin,
+                    OriginalOpResultsIt originalEnd, NewOpResultsIt newBegin,
+                    NewOpResultsIt newEnd, Block& excludedBlock) {
+  assert(std::distance(originalBegin, originalEnd) ==
+         std::distance(newBegin, newEnd));
+  auto newIt = newBegin;
+  for (auto originalIt = originalBegin; originalIt != originalEnd;
+       ++originalIt, ++newIt) {
+    originalIt->replaceUsesWithIf(*newIt, [&excludedBlock](OpOperand& operand) {
+      return !isAncestorOfBlock(operand.getOwner(), &excludedBlock);
+    });
+  }
+}
+
+func::FuncOp createFunctionFromOps(iterator_range<Block::iterator> opRange,
```

Comment:
ditto, brief description would help (explaining how the inputs and outputs are wired together).